### DISCUSSION TITLE
Tracking the local LP gain at each branching decision

### DISCUSSION THREAD

**Discussion by cwfparsonson** [ds4dm/ml4co-competition#93] [Q&A] [answered]
Hi,

I'd like to be able to track the LP gain (i.e. the difference between the relaxed LP solution objective value of the parent node where branching took place and the child node resulting from the branching decision) made after each branching decision. I believe this is equivalent to the multiplicative values used to calculate the strong branching score (but I do not want to have to extract the strong branching scores of all variables to retrospectively calculate this gain for one decision). 

From the `pyscipopt` documentation I can see how to do this if we assume the parent node of the current node was the node we were at when making the last branching decision, but it is not clear to me if it is possible to access the node statistics of the child node which was created from the branching decision if we have moved to a different sub-tree.

I.e. if we make a branching decision and the ecole environment is updated accordingly, if we assume the node selection policy has selected one of the new children, then we can calculate the local LP gain resulting from this branching decision as:

```python
m = env.model.as_pyscipopt()

# new node dual
curr_node = m.getCurrentNode()
curr_node_lb = curr_node.getLowerbound()

# previous node dual
parent_node = curr_node.getParent()
parent_node_lb = parent_node.getLowerbound()

# calc local LP gain resulting from branching decision
lp_gain = parent_node_lb - curr_node_lb
```

But if we have moved to a different sub-tree from the last branching decision and so the parent node of the current node is no longer the previous node we were at, how do we query or track what the LP gain was from the previous branching decision?

**Comment by gasse** [comment, marked as answer]
Hi @cwfparsonson ,

Unfortunately I don't see an easy way to do this with SCIP. What I would do is collect the following information in a list, for all nodes when they are processed:
 - current node's lower bound (local LP value)
 - current node's number
 - parent node's number

You should then be able to retrospectively reconstruct the whole B&B tree, and compute a posteriori the LP gain of each processed node.

Hope that helps,
Maxime

**Reply by cwfparsonson** [reply]
Thanks @gasse, that's exactly what I ended up doing and it seems to be working to track local LP gain!

One thing which I would like to do, which is related to a previous answer you gave me (discussion: https://github.com/ds4dm/ml4co-competition/discussions/62), is to use this to retrospectively calculate the strong branching score of the previous branching decision. 

From page 3 of the Achterberg thesis (https://depositonce.tu-berlin.de/handle/11303/1931 ), I understand the 2 different ways of combining score for the branch children which would be created to get a score for each branching candidate, and that in the case of strong branching, q^{+} and q^{-} are the LP gains of each respective child. The thesis says that SCIP uses the product score function, score(q^{+}, q^{-}) = max(q^{+}, eps) \times max(q^{-}, eps), to calculate the score for each child, where eps=10^{-6}. However, I cannot use this to retrospectively calculate the correct SCIP score which SCIP predicts, and I've tried the linear score function and different values of eps and mu without success. 

Having looked at the SCIP code on line 2190 (https://www.scipopt.org/doc/html/branch_8c_source.php#l02190), I was wondering if the way SCIP calculates strong branching scores might be a bit more complicated than in the Achterberg thesis, because it seems to make use of variable branching factors, quotient functions etc to calculate the score. 

Am I missing something which would let me easily use the local LP gain I can now track to calculate the strong branching score, or would it be a bit more complicated to reproduce the SCIP strong branching score than the Achterberg thesis describes?

**Reply by ambros-gleixner** [reply]
The variable branching factors are something that users can set manually and should be uniformly 1.0 otherwise.

Is the reason maybe that SCIP does not use pure strong branching by default, but hybrid reliability branching? In hybrid reliability branching, the score is computed from a combination of LP dual gains and other statistics involving propagation, conflicts, etc.

**Reply by cwfparsonson** [reply]
To get the strong branching scores (which I'm trying to retrospectively calculate with the above LP gain tracking implementation) I just use the Ecole StrongBranchingScores observation as demonstrated in the Ecole example notebook for imitating strong branching. Are these strong or reliability branching scores?

**Reply by gasse** [reply]
Hi @cwfparsonson ,

The StrongBranchingScores observation produces strong branching scores, alike to the ones computed by SCIP's [`fullstrong`](https://www.scipopt.org/doc/html/branch__fullstrong_8h.php) and [`vanillafullstrong`](https://www.scipopt.org/doc/html/branch__vanillafullstrong_8h.php) branching rules. For the exact formula I refer you to the following places:
https://www.scipopt.org/doc/html/branch_8c_source.php#l02250
https://www.scipopt.org/doc/html/branch__vanillafullstrong_8c_source.php#l00232 (l.199-232)

By default, SCIP will compute SB scores from the product of the gains in the left and right child.
https://www.scipopt.org/doc/html/branch_8c_source.php#l02221

There are a few edge cases though, such as when one of the child nodes is infeasible, or when both are infeasible. Also a few additional things can be triggered when the children nodes are processed (conflict analysis, propagation), or simply a feasible integer solution may have been found in-between, resulting in the pruning of one (or both) children nodes, and therefore those nodes will never be processed. So you should be able to retrospectively reconstruct SB scores approximately, but I'm afraid you can't expect to reconstruct the exact same SB scores are those computed on the fly by, say, [`vanillafullstrong`](https://www.scipopt.org/doc/html/branch__vanillafullstrong_8h.php).

Hope that helps,

Best,
Maxime

**Reply by cwfparsonson** [reply]
Thank you @gasse, that makes perfect sense!
